core/src/main/java/org/elasticsearch/common/geo/builders/MultiPointBuilder.java

```diff
@@ -61,4 +65,39 @@ public Shape build() {
     public GeoShapeType type() {
         return TYPE;
     }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(points);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        MultiPointBuilder other = (MultiPointBuilder) obj;
+        return Objects.equals(points, other.points);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(points.size());
+        for (Coordinate point : points) {
+            writeCoordinateTo(point, out);
+        }
+    }
+
+    @Override
+    public MultiPointBuilder readFrom(StreamInput in) throws IOException {
+        MultiPointBuilder multiPointBuilder = new MultiPointBuilder();
+        int size = in.readVInt();
+        for (int i=0; i < size; i++) {
+            multiPointBuilder.point(readCoordinateFrom(in));
+        }
+        return multiPointBuilder;
```

Comment:
Yes, seems like we didn't cover this in the tests yet, validateMultiPointNode() throws a parse exception here. Since we can't rely on the shape beeing parsed from json any more, we should probably do the same in the java api, but I'm not completely sure where. Doing it inside the serialization method seems weird, we could call a validation method first thing in build() instead but that would mean we only throw error on the shard. In the query refactoring we rejected invalid options in the constructors and setters if possible, but here it would require we always set the whole points-array in one step to be able to check size>0. I'm not sure yet where it would be to catch this but will look into it again.